sklearn/pipeline.py

```diff
@@ -221,22 +227,58 @@ def _validate_steps(self):
                             "'%s' (type %s) doesn't"
                             % (estimator, type(estimator)))
 
+        return deepcopy(self._steps)
+
     @property
     def _estimator_type(self):
-        return self.steps[-1][1]._estimator_type
+        return self._steps[-1][1]._estimator_type
+
+    @property
+    def steps(self):
+        if hasattr(self, 'steps_'):
+            warnings.warn("In the future, 'steps' will be the steps used to"
+                          " instantiate the pipeline. To get the fitted"
+                          " steps, use 'steps_' instead.", FutureWarning)
+            return self.steps_
```

Comment:
which suggests that during deprecation, `steps` and `steps_` need to be literally the same object, or else that the new behaviour has to be opt-in by a constructor parameter `clone_steps=False` which becomes `True` after two versions, and then the parameter is deprecated over two versions. Let's try to avoid it, but we need to be sure that old code working with pipelines, even if non-standard, still works.